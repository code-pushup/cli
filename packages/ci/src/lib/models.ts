import type { SourceFileIssue } from './issues';
import type { MonorepoTool } from './monorepo';

/**
 * Customization options for {@link runInCI}
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#options
 */
export type Options = {
  monorepo?: boolean | MonorepoTool;
  projects?: string[] | null;
  task?: string;
  bin?: string;
  config?: string | null;
  directory?: string;
  silent?: boolean;
  debug?: boolean;
  detectNewIssues?: boolean;
  logger?: Logger;
};

/**
 * {@link Options} with filled-in defaults.
 */
export type Settings = Required<Options>;

/**
 * Branches/commits for {@link runInCI}
 */
export type GitRefs = {
  head: GitBranch;
  base?: GitBranch;
};

/**
 * API client for {@link runInCI} - adapter for CI provider
 * @see https://github.com/code-pushup/cli/tree/main/packages/ci#provider-api-client
 */
export type ProviderAPIClient = {
  maxCommentChars: number;
  downloadReportArtifact?: () => Promise<string | null>;
  listComments: () => Promise<Comment[]>;
  updateComment: (id: number, body: string) => Promise<Comment>;
  createComment: (body: string) => Promise<Comment>;
};

/**
 * PR comment from {@link ProviderAPIClient}
 */
export type Comment = {
  id: number;
  body: string;
  url: string;
};

/**
 * Git ref (e.g. 'main') and commit sha (e.g. '9d7568265bd5b17d1bb51a6eb6407e0d387058d2')
 */
export type GitBranch = {
  ref: string;
  sha: string;
};

/**
 * Logger instance (e.g. `console`) for reporting progress and problems
 */
export type Logger = {
  error: (message: string) => void;
  warn: (message: string) => void;
  info: (message: string) => void;
  debug: (message: string) => void;
};

/**
 * Resolved return value of {@link runInCI}
 */
export type RunResult = StandaloneRunResult | MonorepoRunResult;

/**
 * Resolved return value of {@link runInCI} in standalone mode
 */
export type StandaloneRunResult = Omit<ProjectRunResult, 'name'> & {
  mode: 'standalone';
  commentId?: number;
};

/**
 * Resolved return value of {@link runInCI} in monorepo mode
 */
export type MonorepoRunResult = {
  mode: 'monorepo';
  projects: ProjectRunResult[];
  commentId?: number;
  diffArtifact?: ArtifactData;
};

/**
 * Result of {@link runInCI} for a single project
 */
export type ProjectRunResult = {
  name: string;
  artifacts: {
    report: ArtifactData;
    diff?: ArtifactData;
  };
  newIssues?: SourceFileIssue[];
};

/**
 * Paths to output files from {@link runInCI}, for uploading as job artifact
 */
export type ArtifactData = {
  rootDir: string;
  files: string[];
};
