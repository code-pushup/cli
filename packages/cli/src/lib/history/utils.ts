import { HistoryOptions } from '@code-pushup/core';
import { getHashFromTag, isSemver } from '@code-pushup/utils';
import { HistoryCliOptions } from './history.model';

export async function normalizeHashOptions(
  processArgs: HistoryCliOptions & HistoryOptions,
): Promise<HistoryCliOptions & HistoryOptions> {
  const {
    onlySemverTags,
    // overwritten
    maxCount,
    ...opt
  } = processArgs;

  // eslint-disable-next-line functional/no-let, prefer-const
  let { from, to, ...processOptions } = opt;
  // if no semver filter is used resolve hash of tags, as hashes are used to collect history
  if (!onlySemverTags) {
    if (from && isSemver(from)) {
      const { hash } = await getHashFromTag(from);
      from = hash;
    }
    if (to && isSemver(to)) {
      const { hash } = await getHashFromTag(to);
      to = hash;
    }
  }

  return {
    ...processOptions,
    onlySemverTags,
    maxCount: maxCount && maxCount > 0 ? maxCount : undefined,
    from,
    to,
  };
}
