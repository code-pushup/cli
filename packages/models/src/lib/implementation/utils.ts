import type { CategoryConfig } from '../category-config';
import type { PluginConfig } from '../plugin-config';
import type { PluginReport } from '../report';

/**
 * Regular expression to validate a slug for categories, plugins and audits.
 * - audit (e.g. 'max-lines')
 * - category (e.g. 'performance')
 * Also validates ``and ` `
 */
export const slugRegex = /^[a-z\d]+(?:-[a-z\d]+)*$/;

/**
 * Regular expression to validate a filename.
 */
export const filenameRegex = /^(?!.*[ \\/:*?"<>|]).+$/;

/**
 * helper function to validate string arrays
 *
 * @param strings
 */
export function hasDuplicateStrings(strings: string[]): string[] | false {
  const sortedStrings = [...strings].sort();
  const duplStrings = sortedStrings.filter(
    (item, index) => index !== 0 && item === sortedStrings[index - 1],
  );

  return duplStrings.length === 0 ? false : [...new Set(duplStrings)];
}

/**
 * helper function to validate string arrays
 *
 * @param toCheck
 * @param existing
 */
export function hasMissingStrings(
  toCheck: string[],
  existing: string[],
): string[] | false {
  const nonExisting = toCheck.filter(s => !existing.includes(s));
  return nonExisting.length === 0 ? false : nonExisting;
}

/**
 * helper for error items
 */
export function errorItems(
  items: string[] | false,
  transform: (itemArr: string[]) => string = itemArr => itemArr.join(', '),
): string {
  return transform(items || []);
}

export function exists<T>(value: T): value is NonNullable<T> {
  return value != null;
}

/**
 * Get category references that do not point to any audit or group
 * @param categories
 * @param plugins
 * @returns Array of missing references.
 */
export function getMissingRefsForCategories(
  categories: CategoryConfig[],
  plugins: PluginConfig[] | PluginReport[],
) {
  if (categories.length === 0) {
    return false;
  }

  const auditRefsFromCategory = categories.flatMap(({ refs }) =>
    refs
      .filter(({ type }) => type === 'audit')
      .map(({ plugin, slug }) => `${plugin}/${slug}`),
  );
  const auditRefsFromPlugins = plugins.flatMap(({ audits, slug: pluginSlug }) =>
    audits.map(({ slug }) => `${pluginSlug}/${slug}`),
  );
  const missingAuditRefs = hasMissingStrings(
    auditRefsFromCategory,
    auditRefsFromPlugins,
  );

  const groupRefsFromCategory = categories.flatMap(({ refs }) =>
    refs
      .filter(({ type }) => type === 'group')
      .map(({ plugin, slug }) => `${plugin}#${slug} (group)`),
  );
  const groupRefsFromPlugins = plugins.flatMap(({ groups, slug: pluginSlug }) =>
    Array.isArray(groups)
      ? groups.map(({ slug }) => `${pluginSlug}#${slug} (group)`)
      : [],
  );
  const missingGroupRefs = hasMissingStrings(
    groupRefsFromCategory,
    groupRefsFromPlugins,
  );

  const missingRefs = [missingAuditRefs, missingGroupRefs]
    .filter((refs): refs is string[] => Array.isArray(refs) && refs.length > 0)
    .flat();

  return missingRefs.length > 0 ? missingRefs : false;
}

export function missingRefsForCategoriesErrorMsg(
  categories: CategoryConfig[],
  plugins: PluginConfig[] | PluginReport[],
) {
  const missingRefs = getMissingRefsForCategories(categories, plugins);
  return `The following category references need to point to an audit or group: ${errorItems(
    missingRefs,
  )}`;
}
