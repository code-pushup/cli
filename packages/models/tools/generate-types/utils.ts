import { mkdir, stat, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import { ZodTypeAny } from 'zod';
import { createTypeAlias, printNode, zodToTs } from 'zod-to-ts';

export function generateTypesString(
  schemas: Record<string, ZodTypeAny>,
): string {
  return Object.entries(schemas)
    .map(([identifier, schema]) => {
      const typeName = schemaNameToTypeName(identifier);
      const { node } = zodToTs(schema, typeName);
      const typeAlias = createTypeAlias(node, typeName);
      return `export ${printNode(typeAlias)}`;
    })
    .join('\n');
}

export function schemaNameToTypeName(schemaName: string): string {
  const name = schemaName.replace(/Schema/, '');
  return name[0].toUpperCase() + name.slice(1);
}

export function generateMdFile(typesString: string): string {
  return [
    '# Code PushUp config file reference',
    'The `code-pushup.config.(ts|mjs|js)` file should conform to the following type definition:',
    '```ts',
    typesString,
    '```',
  ].join('\n');
}

export async function safeWriteFile(path: string, content: string) {
  const dir = dirname(path) as string;
  const stats = await stat(dir);
  if (!stats.isDirectory()) {
    await mkdir(dir);
  }
  await writeFile(path, content, { encoding: 'utf8' });
}
