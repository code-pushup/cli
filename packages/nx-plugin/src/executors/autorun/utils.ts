import { globalConfig, persistConfig, uploadConfig } from '../internal/config';
import { NormalizedExecutorContext } from '../internal/context';
import {
  AutorunCommandExecutorOnlyOptions,
  AutorunCommandExecutorOptions,
} from './schema';

export function parseAutorunExecutorOnlyOptions(
  options: Partial<AutorunCommandExecutorOnlyOptions>,
): AutorunCommandExecutorOnlyOptions {
  const { projectPrefix, dryRun, onlyPlugins } = options;
  return {
    ...(projectPrefix && { projectPrefix }),
    ...(dryRun != null && { dryRun }),
    ...(onlyPlugins && { onlyPlugins }),
  };
}

export function parseAutorunExecutorOptions(
  options: Partial<AutorunCommandExecutorOptions>,
  normalizedContext: NormalizedExecutorContext,
): AutorunCommandExecutorOptions {
  const { projectPrefix, persist, upload } = options;
  return {
    ...parseAutorunExecutorOnlyOptions(options),
    ...globalConfig(options, normalizedContext),
    persist: persistConfig({ projectPrefix, ...persist }, normalizedContext),
    upload: uploadConfig({ projectPrefix, ...upload }, normalizedContext),
  };
}
