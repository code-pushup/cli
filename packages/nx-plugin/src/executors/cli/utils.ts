import {
  globalConfig,
  persistConfig,
  uploadConfig,
} from '../internal/config.js';
import type { NormalizedExecutorContext } from '../internal/context.js';
import type {
  AutorunCommandExecutorOnlyOptions,
  AutorunCommandExecutorOptions,
} from './schema.js';

export function parseAutorunExecutorOnlyOptions(
  options: Partial<AutorunCommandExecutorOnlyOptions>,
): AutorunCommandExecutorOnlyOptions {
  const { projectPrefix, dryRun, onlyPlugins } = options;
  return {
    ...(projectPrefix && { projectPrefix }),
    ...(dryRun != null && { dryRun }),
    ...(onlyPlugins && { onlyPlugins }),
  };
}

export function parseAutorunExecutorOptions(
  options: Partial<AutorunCommandExecutorOptions>,
  normalizedContext: NormalizedExecutorContext,
): AutorunCommandExecutorOptions {
  const { projectPrefix, persist, upload, command } = options;
  const needsUploadParams =
    command === 'upload' || command === 'autorun' || command === undefined;
  const uploadCfg = uploadConfig(
    { projectPrefix, ...upload },
    normalizedContext,
  );
  const hasApiToken = uploadCfg?.apiKey != null;
  return {
    ...parseAutorunExecutorOnlyOptions(options),
    ...globalConfig(options, normalizedContext),
    persist: persistConfig({ projectPrefix, ...persist }, normalizedContext),
    // @TODO This is a hack to avoid validation errors of upload config for commands that dont need it.
    // Fix: use utils and execute the core logic directly
    // Blocked by Nx plugins can't compile to es6
    ...(needsUploadParams && hasApiToken ? { upload: uploadCfg } : {}),
  };
}

/**
 * Deeply merges executor options.
 *
 * @param targetOptions - The original options from the target configuration.
 * @param cliOptions - The options from Nx, combining target options and CLI arguments.
 * @returns A new object with deeply merged properties.
 *
 * Nx performs a shallow merge by default, where command-line arguments can override entire objects
 * (e.g., `--persist.filename` replaces the entire `persist` object).
 * This function ensures that nested properties are deeply merged,
 * preserving the original target options where CLI arguments are not provided.
 */
export function mergeExecutorOptions(
  targetOptions: Partial<AutorunCommandExecutorOptions>,
  cliOptions: Partial<AutorunCommandExecutorOptions>,
): AutorunCommandExecutorOptions {
  return {
    ...targetOptions,
    ...cliOptions,
    ...(targetOptions?.persist || cliOptions?.persist
      ? {
          persist: {
            ...targetOptions?.persist,
            ...cliOptions?.persist,
          },
        }
      : {}),
    ...(targetOptions?.upload || cliOptions?.upload
      ? {
          upload: {
            ...targetOptions?.upload,
            ...cliOptions?.upload,
          },
        }
      : {}),
  };
}
