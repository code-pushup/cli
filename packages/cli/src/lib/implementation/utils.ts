import { toArray } from '@code-pushup/utils';

// log error and flush stdout so that Yargs doesn't suppress it
// related issue: https://github.com/yargs/yargs/issues/2118
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function logErrorBeforeThrow<T extends (...args: any[]) => any>(
  fn: T,
): T {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (async (...args: any[]) => {
    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument
      return await fn(...args);
    } catch (error) {
      console.error(error);
      await new Promise(resolve => process.stdout.write('', resolve));
      throw error;
    }
  }) as T;
}

export function coerceArray<T extends string>(param: T | T[] = []): T[] {
  return [
    ...new Set(toArray(param).flatMap((f: T) => f.split(',') as T[]) || []),
  ];
}
