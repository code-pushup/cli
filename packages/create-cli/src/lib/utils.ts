import { readFile, rm, stat, writeFile } from 'node:fs/promises';
import { CODE_PUSHUP_UNICODE_LOGO } from '@code-pushup/utils';
import {
  NX_JSON_CONTENT,
  NX_JSON_FILENAME,
  PROJECT_JSON_CONTENT,
  PROJECT_JSON_FILENAME,
  PROJECT_NAME,
} from './constants.js';

export type SetupResult = {
  filename: string;
  teardown: boolean;
};

export async function setupFile(
  filename: string,
  content = '',
): Promise<SetupResult> {
  const setupResult: SetupResult = {
    filename,
    teardown: false,
  };

  try {
    const stats = await stat(filename);
    if (!stats.isFile()) {
      await writeFile(filename, content);
    }
  } catch (error) {
    if (
      error instanceof Error &&
      error.message.includes('no such file or directory')
    ) {
      await writeFile(filename, content);
      return {
        ...setupResult,
        teardown: true,
      };
    } else {
      console.error(error);
    }
  }

  return setupResult;
}

export function parseNxProcessOutput(output: string) {
  return output.trim().replace('NX', CODE_PUSHUP_UNICODE_LOGO);
}

export async function setupNxContext(): Promise<{
  nxJsonTeardown: boolean;
  projectJsonTeardown: boolean;
  projectName: string;
}> {
  const { teardown: nxJsonTeardown } = await setupFile(
    NX_JSON_FILENAME,
    NX_JSON_CONTENT,
  );
  const { teardown: projectJsonTeardown } = await setupFile(
    PROJECT_JSON_FILENAME,
    PROJECT_JSON_CONTENT,
  );

  const projectJsonContent = await readFile(PROJECT_JSON_FILENAME, 'utf8');
  const { name = PROJECT_NAME } = JSON.parse(projectJsonContent) as {
    name: string;
  };

  return {
    nxJsonTeardown,
    projectJsonTeardown,
    projectName: name,
  };
}

export async function teardownNxContext({
  nxJsonTeardown,
  projectJsonTeardown,
}: {
  nxJsonTeardown: boolean;
  projectJsonTeardown: boolean;
}) {
  const filesToDelete = [
    ...(nxJsonTeardown ? [NX_JSON_FILENAME] : []),
    ...(projectJsonTeardown ? [PROJECT_JSON_FILENAME] : []),
  ];
  await Promise.all(filesToDelete.map(file => rm(file)));
}
