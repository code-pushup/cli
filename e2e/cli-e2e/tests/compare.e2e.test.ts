import { simpleGit } from 'simple-git';
import { ReportsDiff } from '@code-pushup/models';
import { cleanTestFolder } from '@code-pushup/test-setup';
import { executeProcess, readJsonFile } from '@code-pushup/utils';

describe('CLI compare', () => {
  const git = simpleGit();

  beforeEach(async () => {
    if (await git.diff(['--', 'examples/react-todos-app'])) {
      throw new Error(
        'Unstaged changes found in examples/react-todos-app, please stage or commit them to prevent E2E tests interfering',
      );
    }
    await cleanTestFolder('tmp/e2e');
    await executeProcess({
      command: 'code-pushup',
      args: ['collect', '--persist.filename=source-report'],
      cwd: 'examples/react-todos-app',
    });
    await executeProcess({
      command: 'npx',
      args: ['eslint', '--fix', 'src', '--ext=js,jsx'],
      cwd: 'examples/react-todos-app',
    });
    await executeProcess({
      command: 'code-pushup',
      args: ['collect', '--persist.filename=target-report'],
      cwd: 'examples/react-todos-app',
    });
  });

  afterEach(async () => {
    await git.checkout(['--', 'examples/react-todos-app']);
    await cleanTestFolder('tmp/e2e');
  });

  it('should compare report.json files and create report-diff.json', async () => {
    await executeProcess({
      command: 'code-pushup',
      args: [
        'compare',
        '--before=../../tmp/e2e/react-todos-app/source-report.json',
        '--after=../../tmp/e2e/react-todos-app/target-report.json',
      ],
      cwd: 'examples/react-todos-app',
    });

    const reportsDiff = await readJsonFile<ReportsDiff>(
      'tmp/e2e/react-todos-app/report-diff.json',
    );
    expect(reportsDiff).toMatchSnapshot({
      commits: expect.any(Object),
      date: expect.any(String),
      duration: expect.any(Number),
    });
  });
});
