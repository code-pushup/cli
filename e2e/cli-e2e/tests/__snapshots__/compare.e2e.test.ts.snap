// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI compare > should compare report.json files and create report-diff.json and report-diff.md 1`] = `
{
  "audits": {
    "added": [],
    "changed": [
      {
        "displayValues": {
          "after": "passed",
          "before": "1 warning",
        },
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "scores": {
          "after": 1,
          "before": 0,
          "diff": 1,
        },
        "slug": "arrow-body-style",
        "title": "Require braces around arrow function bodies",
        "values": {
          "after": 0,
          "before": 1,
          "diff": -1,
        },
      },
      {
        "displayValues": {
          "after": "passed",
          "before": "3 warnings",
        },
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "scores": {
          "after": 1,
          "before": 0,
          "diff": 1,
        },
        "slug": "object-shorthand",
        "title": "Require or disallow method and property shorthand syntax for object literals",
        "values": {
          "after": 0,
          "before": 3,
          "diff": -3,
        },
      },
      {
        "displayValues": {
          "after": "passed",
          "before": "1 warning",
        },
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "scores": {
          "after": 1,
          "before": 0,
          "diff": 1,
        },
        "slug": "prefer-const",
        "title": "Require \`const\` declarations for variables that are never reassigned after declared",
        "values": {
          "after": 0,
          "before": 1,
          "diff": -1,
        },
      },
    ],
    "removed": [],
    "unchanged": [
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-cond-assign",
        "title": "Disallow assignment operators in conditional expressions",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-const-assign",
        "title": "Disallow reassigning \`const\` variables",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-debugger",
        "title": "Disallow the use of \`debugger\`",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-invalid-regexp",
        "title": "Disallow invalid regular expression strings in \`RegExp\` constructors",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-undef",
        "title": "Disallow the use of undeclared variables unless mentioned in \`/*global */\` comments",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-unreachable-loop",
        "title": "Disallow loops with a body that allows only one iteration",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-unsafe-negation",
        "title": "Disallow negating the left operand of relational operators",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-unsafe-optional-chaining",
        "title": "Disallow use of optional chaining in contexts where the \`undefined\` value is not allowed",
        "value": 0,
      },
      {
        "displayValue": "1 warning",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0,
        "slug": "no-unused-vars",
        "title": "Disallow unused variables",
        "value": 1,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "use-isnan",
        "title": "Require calls to \`isNaN()\` when checking for \`NaN\`",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "valid-typeof",
        "title": "Enforce comparing \`typeof\` expressions against valid strings",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "camelcase",
        "title": "Enforce camelcase naming convention",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "curly",
        "title": "Enforce consistent brace style for all control statements",
        "value": 0,
      },
      {
        "displayValue": "1 warning",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0,
        "slug": "eqeqeq",
        "title": "Require the use of \`===\` and \`!==\`",
        "value": 1,
      },
      {
        "displayValue": "1 warning",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0,
        "slug": "max-lines-per-function",
        "title": "Enforce a maximum number of lines of code in a function",
        "value": 1,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "max-lines",
        "title": "Enforce a maximum number of lines per file",
        "value": 0,
      },
      {
        "displayValue": "3 warnings",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0,
        "slug": "no-shadow",
        "title": "Disallow variable declarations from shadowing variables declared in the outer scope",
        "value": 3,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "no-var",
        "title": "Require \`let\` or \`const\` instead of \`var\`",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "prefer-arrow-callback",
        "title": "Require using arrow functions for callbacks",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "prefer-object-spread",
        "title": "Disallow using Object.assign with an object literal as the first argument and prefer the use of object spread instead",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "yoda",
        "title": "Require or disallow "Yoda" conditions",
        "value": 0,
      },
      {
        "displayValue": "1 warning",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0,
        "slug": "react-jsx-key",
        "title": "Disallow missing \`key\` props in iterators/collection literals",
        "value": 1,
      },
      {
        "displayValue": "6 warnings",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0,
        "slug": "react-prop-types",
        "title": "Disallow missing props validation in a React component definition",
        "value": 6,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-react-in-jsx-scope",
        "title": "Disallow missing React when using JSX",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-hooks-rules-of-hooks",
        "title": "enforces the Rules of Hooks",
        "value": 0,
      },
      {
        "displayValue": "2 warnings",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0,
        "slug": "react-hooks-exhaustive-deps",
        "title": "verifies the list of dependencies for Hooks like useEffect and similar",
        "value": 2,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-display-name",
        "title": "Disallow missing displayName in a React component definition",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-jsx-no-comment-textnodes",
        "title": "Disallow comments from being inserted as text nodes",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-jsx-no-duplicate-props",
        "title": "Disallow duplicate properties in JSX",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-jsx-no-target-blank",
        "title": "Disallow \`target="_blank"\` attribute without \`rel="noreferrer"\`",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-jsx-no-undef",
        "title": "Disallow undeclared variables in JSX",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-jsx-uses-react",
        "title": "Disallow React to be incorrectly marked as unused",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-jsx-uses-vars",
        "title": "Disallow variables used in JSX to be incorrectly marked as unused",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-children-prop",
        "title": "Disallow passing of children as props",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-danger-with-children",
        "title": "Disallow when a DOM element is using both children and dangerouslySetInnerHTML",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-deprecated",
        "title": "Disallow usage of deprecated methods",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-direct-mutation-state",
        "title": "Disallow direct mutation of this.state",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-find-dom-node",
        "title": "Disallow usage of findDOMNode",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-is-mounted",
        "title": "Disallow usage of isMounted",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-render-return-value",
        "title": "Disallow usage of the return value of ReactDOM.render",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-string-refs",
        "title": "Disallow using string references",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-unescaped-entities",
        "title": "Disallow unescaped HTML entities from appearing in markup",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-no-unknown-property",
        "title": "Disallow usage of unknown DOM property",
        "value": 0,
      },
      {
        "displayValue": "passed",
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 1,
        "slug": "react-require-render-return",
        "title": "Enforce ES5 or ES6 class for returning value in render function",
        "value": 0,
      },
    ],
  },
  "categories": {
    "added": [],
    "changed": [
      {
        "scores": {
          "after": 0.7692307692307693,
          "before": 0.5384615384615384,
          "diff": 0.23076923076923084,
        },
        "slug": "code-style",
        "title": "Code style",
      },
    ],
    "removed": [],
    "unchanged": [
      {
        "score": 0.6842105263157895,
        "slug": "bug-prevention",
        "title": "Bug prevention",
      },
    ],
  },
  "commits": Any<Object>,
  "date": Any<String>,
  "duration": Any<Number>,
  "groups": {
    "added": [],
    "changed": [
      {
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "scores": {
          "after": 0.7142857142857143,
          "before": 0.5,
          "diff": 0.2142857142857143,
        },
        "slug": "suggestions",
        "title": "Suggestions",
      },
    ],
    "removed": [],
    "unchanged": [
      {
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0.9166666666666666,
        "slug": "problems",
        "title": "Problems",
      },
      {
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0.9090909090909091,
        "slug": "react-best-practices",
        "title": "Best Practices (react)",
      },
      {
        "plugin": {
          "slug": "eslint",
          "title": "ESLint",
        },
        "score": 0.9,
        "slug": "react-possible-errors",
        "title": "Possible Errors (react)",
      },
    ],
  },
  "packageName": "@code-pushup/core",
  "version": Any<String>,
}
`;
